pkg/pushsync/pushsync_test.go

```diff
@@ -2,4 +2,336 @@
+	"github.com/ethersphere/bee/pkg/p2p/streamtest"
+	mockstate "github.com/ethersphere/bee/pkg/statestore/mock"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology/full"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// TestSencChunk tests that a chunk that is uploaded to localstore is sent to the appropriate closest peer.
+func TestSendChunk(t *testing.T) {
+	logger := logging.New(ioutil.Discard, 0)
+
+	// chunk data to upload
+	chunkAddress := swarm.MustParseHexAddress("7000000000000000000000000000000000000000000000000000000000000000")
+	chunkData := []byte("1234")
+
+	// create a pivot node and a cluster of nodes
+	pivotNode := swarm.MustParseHexAddress("0000000000000000000000000000000000000000000000000000000000000000") // base is 0000
+	connectedPeers := []p2p.Peer{
+		{
+			Address: swarm.MustParseHexAddress("8000000000000000000000000000000000000000000000000000000000000000"), // binary 1000 -> po 0
+		},
+		{
+			Address: swarm.MustParseHexAddress("4000000000000000000000000000000000000000000000000000000000000000"), // binary 0100 -> po 1
+		},
+		{
+			Address: swarm.MustParseHexAddress("6000000000000000000000000000000000000000000000000000000000000000"), // binary 0110 -> po 1
+		},
+	}
+	closestPeer := connectedPeers[2].Address
+
+	// mock a connectivity between the nodes
+	p2ps := p2pmock.New(p2pmock.WithConnectFunc(func(ctx context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+		return pivotNode, nil
+	}), p2pmock.WithPeersFunc(func() []p2p.Peer {
+		return connectedPeers
+	}))
+
+	// Create a full connectivity between the peers
+	discovery := mock.NewDiscovery()
+	statestore := mockstate.NewStateStore()
+	ab := addressbook.New(statestore)
+	fullDriver := full.New(discovery, ab, p2ps, logger, pivotNode)
+
+	storer, err := localstore.New("", pivotNode.Bytes(), nil, logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// setup the stream recorder to record stream data
+	recorder := streamtest.New(
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
+			return func(context.Context, p2p.Peer, p2p.Stream) error {
+				// dont call any handlers
+				return nil
+			}
+		}),
+	)
+
+	// instantiate a pushsync instance
+	ps := pushsync.New(pushsync.Options{
+		Streamer:      recorder,
+		Logger:        logger,
+		ClosestPeerer: fullDriver,
+		Storer:        storer,
+	})
+	defer ps.Close()
+	recorder.SetProtocols(ps.Protocol())
+
+	// upload the chunk to the pivot node
+	_, err = storer.Put(context.Background(), storage.ModePutUpload, swarm.NewChunk(chunkAddress, chunkData))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var records []*streamtest.Record
+
+LOOP:
+	for i := 0; i < 10; i++ {
+		records, _ = recorder.Records(closestPeer, pushsync.ProtocolName, pushsync.ProtocolVersion, pushsync.StreamName)
+		switch l := len(records); l {
+		case 1:
+			break LOOP
+		case 0:
+			time.Sleep(time.Millisecond * 10)
```

Comment:
ewww

Reply:
Is total of 100ms to wait good enough for the slow ci?

Reply:
i will raise it